web3/providers/websocket/websocket_v2.py

```diff
@@ -158,40 +157,55 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
+            # if response is already cached, pop it from cache
+            response = await self._request_processor.pop_raw_response(request_cache_key)
+        else:
+            # else, wait for the desired response, caching all others along the way
+            response = await self._get_response_for_request_id(current_request_id)
 
-        if response_id != current_request_id:
-            request_cache_key = generate_cache_key(current_request_id)
-            if request_cache_key in self._request_processor._raw_response_cache:
-                async with async_lock(self._thread_pool, self._lock):
-                    # if response is already cached, pop it from the cache
-                    response = self._request_processor.pop_raw_response(
-                        request_cache_key
-                    )
-            else:
-                async with async_lock(self._thread_pool, self._lock):
-                    # cache response
-                    self._request_processor.cache_raw_response(response)
-                    response = await asyncio.wait_for(
-                        self._get_response_for_request_id(current_request_id),
-                        self.call_timeout,
-                    )
         return response
 
     async def _get_response_for_request_id(self, request_id: RPCId) -> RPCResponse:
-        response = await self._ws_recv()
-        response_id = response.get("id")
-
-        while response_id != request_id:
-            response = await self._ws_recv()
-            response_id = response.get("id")
-            if response_id != request_id:
-                self._request_processor.cache_raw_response(
-                    response,
-                )
-        return response
+        async def _match_response_id_to_request_id() -> RPCResponse:
+            response_id = None
+            response = None
+            while response_id != request_id:
+                response = await self._ws_recv()
+                response_id = response.get("id")
+
+                if response_id == request_id:
+                    break
+                else:
+                    # cache all responses that are not the desired response
+                    await self._request_processor.cache_raw_response(
+                        response,
+                    )
+                    await asyncio.sleep(0.1)
+
+            return response
+
+        try:
+            # Enters a while loop, looking for a response id match to the request id.
+            # If the provider does not give responses with matching ids, this will
+            # hang forever. The JSON-RPC spec requires that providers respond with
+            # the same id that was sent in the request, but we need to handle these
+            # "bad" cases somewhat gracefully.
+            timeout = (
+                self.call_timeout
+                if self.call_timeout and self.call_timeout <= 20
+                else 20
```

Comment:
Should use the `DEFAULT_PERSISTENT_CONNECTION_TIMEOUT` constant here.
```suggestion
                if self.call_timeout and self.call_timeout <= DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
                else DEFAULT_PERSISTENT_CONNECTION_TIMEOUT
```

Reply:
Isn't this an entirely different bit of logic though? What if we agree to change the default timeout? I think that was the point of hard-coding this here.

Let's say I don't want _each_ call to the provider to take more than, say, 5 seconds. I wonder if I'm subscribed to a lot of subscriptions, if the response to my call won't be bogged down by the latency behind so many subscriptions where it may actually take longer than a call timeout to receive it.

Thoughts there?

Reply:
That's fine then, I have no concerns with using the hardcoded value as is. Definitely would be overkill to create yet another constant just for this logic.

Reply:
Yeah in the same way I think using `self.call_timeout` feels a bit disingenuous to use here too but at least the user can tweak that to make sense if desired. Definitely open to re-thinking this a bit more before ripping off the `beta` tag.